### PR TITLE
FirestoreDb-registered custom converters

### LINKED ISSUE #3255: FireStore: Register converters on client
Currently the FireStore converters are stored inside of a static cache that can only be accessed internally by the library. Can we change it so that the cache is instead on a client and allow the user add converters to this client cache?

My reasoning for this is that currently the only way to add converters for FireStore is by adding attributes to the POCO objects, which I could prefer to avoid. One, it requires me to reference FireStore in a part of our project where I don't feel it is required. Secondly, it may make it difficult when trying to use POCO's from an external library where one does not have access to the code base to add FireStore attributes.

By putting the converter cache on the client and allowing the public to add converters opens up the possibility of writing converters without having to rely on the FireStore attributes.

### PR DESCRIPTION
**Pull request by jskeet**
This PR will eventually fix #3255.

### CODE REVIEW

**Comment by jskeet** [comment]
I'm now talking about this with the Firestore team, and the more I've written about it, the more I'd really quite like it in. Thoughts on overall API design and feature usefulness are welcome as well as the implementation details.

**Comment by jskeet** [comment]
Firestore team is now happy, so we're ready for review. I'll need to update the docs when it's in, but I might wait until I've finished another couple of Firestore serialization features first, and update the docs in one go.

File: apis/Google.Cloud.Firestore/Google.Cloud.Firestore.Benchmarks/ValueSerializerBenchmark.cs [hunk 1]
```diff
@@ -20,36 +20,38 @@ namespace Google.Cloud.Firestore.Benchmarks
 {
     public class ValueSerializerBenchmark
     {
+        private static readonly SerializationContext _serializationContext = new SerializationContext(null);
```
Comment (chrisdunelm) [inline apis/Google.Cloud.Firestore/Google.Cloud.Firestore.Benchmarks/ValueSerializerBenchmark.cs]: `= SerializationContext.Default` ?

File: apis/Google.Cloud.Firestore/Google.Cloud.Firestore/SerializationContext.cs [hunk 2]
```diff
@@ -24,11 +25,22 @@ namespace Google.Cloud.Firestore
     /// </summary>
     internal sealed class SerializationContext
     {
+        // Important note: if this ever contains more state than just the converters,
+        // e.g. a reference to the Database, then WithWarningLogger will need to be adjusted accordingly.
+        
+        private readonly IReadOnlyDictionary<BclType, IFirestoreInternalConverter> _customConverters;
```
Comment (chrisdunelm) [inline apis/Google.Cloud.Firestore/Google.Cloud.Firestore/SerializationContext.cs]: nit: place this *after* the static `Default` property?

File: apis/Google.Cloud.Firestore/Google.Cloud.Firestore.Benchmarks/ValueSerializerBenchmark.cs (hunk 1 above)
Comment (jskeet) [reply apis/Google.Cloud.Firestore/Google.Cloud.Firestore.Benchmarks/ValueSerializerBenchmark.cs]: Good idea. I probably hadn't created SerializationContext.Default when I made that change :)

File: apis/Google.Cloud.Firestore/Google.Cloud.Firestore/SerializationContext.cs (hunk 2 above)
Comment (jskeet) [reply apis/Google.Cloud.Firestore/Google.Cloud.Firestore/SerializationContext.cs]: Ooh yes, definitely.

**Pull request merged**